### PR TITLE
fix String encoding

### LINKED ISSUE #233: String with encoding
There seems to be some inconsistencies with respect to the `encoding` argument on `String` objects.  For example (in py3):
```
>>> import colander
>>> s = colander.String('utf-8')
>>> s.serialize(None, 123)
'123'
>>> s.serialize(None, '123')
b'123'
>>> s.serialize(None, str(123))
b'123'
```
in py2:
```
>>> import colander
>>> s = colander.String('utf-8')
>>> s.serialize(None, 123)
u'123'
>>> s.serialize(None, '123')
'123'
>>> s.serialize(None, unicode(123))
'123'
```
Basically, if you pass something that isn't an instance of one of the python string types you can bypass the encoding step.

The docs say:
> A non-Unicode input value to ``serialize`` is converted to a
> Unicode using the encoding (``unicode(value, encoding)``);
> subsequently the Unicode object is reeencoded to a ``str``
> object using the encoding and returned.

This seems to be a regression as it appears it did work properly in previous versions.

### PR DESCRIPTION
**Pull request by tisdall**
fixes #233

### CODE REVIEW

File: colander/__init__.py [hunk 1]
```diff
@@ -1184,7 +1184,10 @@ def serialize(self, node, appstruct):
                 else:
                     result = text_type(appstruct)
             else:
-                result = text_type(appstruct)
+                if self.encoding:
+                    result = text_type(appstruct).encode(self.encoding)
+                else:
+                    result = text_type(appstruct)
```
Comment (tseaver) [inline colander/__init__.py]: I prefer that the `result = text_type(appstruct)` to be outside the `if`, with no `else`.  E.g.:

```python
    result = text_type(appstruct)
    if self.encoding:
        result = result.encode(self.encoding)
    return result
```

**Comment by tisdall** [comment]
@tseaver - updated as requested...

File: colander/tests/test_colander.py [hunk 2]
```diff
@@ -1557,6 +1557,13 @@ def test_serialize_string_with_high_unresolveable_high_order_chars(self):
         e = invalid_exc(typ.serialize, node, not_utf8)
         self.assertTrue('cannot be serialized' in e.msg)
 
+    def test_serialize_encoding_with_non_string_type(self):
+        utf8 = '123'.encode('utf-8')
```
Comment (tseaver) [inline colander/tests/test_colander.py]: Hmm, this one should start as text and then encode to bytes (Python2 lets us get away with being sloppy, but that doesn't mean we shouldn't be explicit).

Comment (tisdall) [inline colander/tests/test_colander.py]: Sorry, I develop everything in Python 3 and then deal with Python 2 only when I need backwards compatibility.  I know this is already doing what your describing in Py3, but I don't understand how it's not doing this in Py2.  In Py2, should this be `u'123'.encode('utf-8')` to be explicit?

Comment (tisdall) [inline colander/tests/test_colander.py]: Are you saying this should be `utf8 = text_type('123').encode('utf-8')`?

Comment (tseaver) [inline colander/tests/test_colander.py]: That would be the better spelling, yes.

Comment (tseaver) [inline colander/tests/test_colander.py]: Or just b'123', in fact (ASCII and UTF-8 are identical encodings for 7-bit values).

**Comment by tisdall** [comment]
@tseaver - I went with using `text_type('123').encode('utf-8')` so it's totally explicit that the result is a utf-8 encoded string.

**Comment by tisdall** [comment]
@tseaver - should I add a note in the CHANGES.rst?  Is there anything else I need to change to complete this?

**Comment by tseaver** [comment]
A note in `CHANGES.rst` would be great, thanks!

**Comment by tisdall** [comment]
done

**Comment by tseaver** [comment]
Thanks!

**Pull request merged**